### PR TITLE
APS-REVAMP-01

### PR DESCRIPTION
**Pull request by mahendra-aps**
Added actor based projects core,tpp and user.

### CODE REVIEW

File: tpp/log4net.config [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<log4net>
```
Comment (yuhanniu) [inline tpp/log4net.config]: Just wanted to double that this file is required for all core, user, tpp?

File: tpp/tpp.csproj [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <AssemblyName>Token.SDK.Net.Tpp</AssemblyName>
+    <RootNamespace>Tokenio</RootNamespace>
+    <Copyright>Copyright 2019 Token, Inc</Copyright>
+    <Description>C# SDK for interacting with the Token System</Description>
+    <Company>Token, Inc</Company>
+    <Authors>Token, Inc</Authors>
+    <Version>2.0.0-beta1</Version>
```
Comment (yuhanniu) [inline tpp/tpp.csproj]: 1.0.0-beta1

File: tpp/tpp.csproj [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    <Authors>Token, Inc</Authors>
+    <Version>2.0.0-beta1</Version>
+    <PackageLicenseExpression>MIT</PackageLicenseExpression>
+    <PackageIconUrl>https://s3-us-west-2.amazonaws.com/tokenio-assets/1.0.0/logos/token_black.png</PackageIconUrl>
+    <RepositoryUrl>https://github.com/tokenio/sdk-csharp</RepositoryUrl>
+    </PropertyGroup>
+ 
+ 
+  <ItemGroup>
+    <ProjectReference Include="..\core\core.csproj" />
```
Comment (yuhanniu) [inline tpp/tpp.csproj]: Core.csproj

File: user/user.csproj [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <AssemblyName>Token.SDK.Net.User</AssemblyName>
+    <RootNamespace>Tokenio</RootNamespace>
+    <Copyright>Copyright 2019 Token, Inc</Copyright>
+    <Description>C# SDK for interacting with the Token System</Description>
+    <Company>Token, Inc</Company>
+    <Authors>Token, Inc</Authors>
+    <Version>2.0.0-beta1</Version>
```
Comment (yuhanniu) [inline user/user.csproj]: 1.0.0-beta1

File: user/user.csproj [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    <PackageLicenseExpression>MIT</PackageLicenseExpression>
+    <PackageIconUrl>https://s3-us-west-2.amazonaws.com/tokenio-assets/1.0.0/logos/token_black.png</PackageIconUrl>
+    <RepositoryUrl>https://github.com/tokenio/sdk-csharp</RepositoryUrl>
+    </PropertyGroup>
+  
+
+  <ItemGroup>
+    <ProjectReference Include="..\core\core.csproj" />
```
Comment (yuhanniu) [inline user/user.csproj]: Core.csproj

File: user/src/TransferTokenBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+		/// Sets the flag indicating whether a receipt is requested.
+		/// </summary>
+		/// <param name="receiptRequested">receipt requested flag</param>
+		/// <returns>builder</returns>
+		public TransferTokenBuilder SetReceiptRequested(bool receiptRequested) {
+			payload.ReceiptRequested = receiptRequested;
+			return this;
+		}
+
+		public TransferTokenBuilder From(string memberId) {
```
Comment (gianluca-token) [inline user/src/TransferTokenBuilder.cs]: comment?

File: build_proto_revamp.rb [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+file = "#{name}-#{type}-#{version}.jar"
+puts("Downloading #{file} ...")
+
+m2path = ENV["HOME"] + "/.m2/repository/#{path}/#{name}-#{type}/#{version}/#{file}"
+if File.file?(m2path) then
+FileUtils.cp(m2path, file)
+else
+url = "https://token.jfrog.io/token/libs-release/#{path}/#{name}-#{type}/#{version}/#{file}"
+open(file, 'wb') do |file|
+file << open(url, :http_basic_authentication => ['vanika.appperfect', 'App4ever$']).read
```
[the remark this answers is not in the document]
Comment (mahendra-aps) [reply build_proto_revamp.rb]: Done

File: tpp/tpp.csproj (hunk 2 above)
Comment (mahendra-aps) [reply tpp/tpp.csproj]: Getting below error after changing the version to 1.0.0-beta1

System.AggregateException : One or more errors occurred. (Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK."))
---- Grpc.Core.RpcException : Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK.")

File: tpp/tpp.csproj (hunk 3 above)
Comment (mahendra-aps) [reply tpp/tpp.csproj]: Getting below error after changing the version to 1.0.0-beta1

System.AggregateException : One or more errors occurred. (Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK."))
---- Grpc.Core.RpcException : Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK.")

File: user/user.csproj (hunk 4 above)
Comment (mahendra-aps) [reply user/user.csproj]: Getting below error after changing the version to 1.0.0-beta1

System.AggregateException : One or more errors occurred. (Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK."))
---- Grpc.Core.RpcException : Status(StatusCode=FailedPrecondition, Detail="Unsupported client version 1.0.0. Please upgrade your csharp SDK.")

File: core/core.csproj [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <AssemblyName>Token.SDK.Net.Core</AssemblyName>
+    <RootNamespace>Tokenio</RootNamespace>
+    <Copyright>Copyright 2019 Token, Inc</Copyright>
+    <Description>C# SDK for interacting with the Token System</Description>
+    <Company>Token, Inc</Company>
+    <Authors>Token, Inc</Authors>
+    <Version>2.0.0-beta1</Version>
```
[the remark this answers is not in the document]
Comment (mahendra-aps) [reply core/core.csproj]: done

File: sdk-csharp.sln [hunk 9]
```diff
@@ -1,11 +1,19 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 15
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "SDK", "sdk\SDK.csproj", "{7B0E8569-DDCC-4F16-AD62-DB57812B4298}"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Sample", "samples\Sample.csproj", "{AED81F7F-B9A2-41F1-B6CD-53A3AA626A24}"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Test", "tests\Test.csproj", "{23F293C9-A369-4A5A-A643-A80D932ADD7A}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "core", "core\core.csproj", "{8A173027-F8A0-4A24-B843-0349AB5107BA}"
```
[the remark this answers is not in the document]
Comment (mahendra-aps) [reply sdk-csharp.sln]: done

File: sdk-csharp.sln [hunk 10]
```diff
@@ -1,11 +1,19 @@
...
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "tpp", "tpp\tpp.csproj", "{65FADF43-4C2B-4C23-AA1C-BF0C25B207A7}"
```
[the remark this answers is not in the document]
Comment (mahendra-aps) [reply sdk-csharp.sln]: done

File: sdk-csharp.sln [hunk 11]
```diff
@@ -1,11 +1,19 @@
...
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "user", "user\user.csproj", "{08EF9494-B16F-444B-AF58-DFCD43DC1357}"
```
[the remark this answers is not in the document]
Comment (mahendra-aps) [reply sdk-csharp.sln]: done

File: tpp/log4net.config (hunk 1 above)
Comment (mahendra-aps) [reply tpp/log4net.config]: Done.Removed from all core, user, tpp

File: tpp/tpp.csproj (hunk 3 above)
Comment (mahendra-aps) [reply tpp/tpp.csproj]: done

File: user/user.csproj (hunk 5 above)
Comment (mahendra-aps) [reply user/user.csproj]: done

File: user/src/Member.cs [hunk 12]
```diff
@@ -0,0 +1,1462 @@
...
+		/// <param name="limit">max number of records to return</param>
+		/// <returns>a paged list of transfers</returns>
+		public PagedList<Transfer> GetTransfersBlocking(
+				string tokenId,
+				string offset,
+				int limit) {
+			return GetTransfers(tokenId, offset, limit).Result;
+		}
+
+		public Task<PrepareTokenResult> PrepareTransferToken(
```
[the remark this answers is not in the document]
Comment (mahendra-aps) [reply user/src/Member.cs]: done

File: user/src/Member.cs [hunk 13]
```diff
@@ -0,0 +1,1462 @@
...
+		public Task<Token> CreateTransferToken(TokenPayload payload, string tokenRequestId) {
+			return client.CreateTransferToken(payload, tokenRequestId);
+		}
+
+		/// <summary>
+		/// Creates a new transfer token builder from a token request.
+		/// </summary>
+		/// <param name="tokenRequest">token request</param>
+		/// <returns>transfer token builder</returns>
+		public TransferTokenBuilder CreateTransferTokenBlocking(Proto.Common.TokenProtos.TokenRequest tokenRequest) {
```
[the remark this answers is not in the document]
Comment (mahendra-aps) [reply user/src/Member.cs]: Done. Changed to CreateTransferTokenBuilder

File: user/src/TransferTokenBuilder.cs (hunk 6 above)
Comment (mahendra-aps) [reply user/src/TransferTokenBuilder.cs]: done

**Comment by mahendra-aps** [comment]
> What you’ve written looks good, but I think you are missing two things:
> 
> 1. I don’t see new sample code and tests. There should be separate sample code for the tpp and user SDKs as well as separate tests for each. See user-sample and tpp-sample in the java SDK as examples.
> 2. I think initiateAccountLinking is missing? I see you copied the browser interfaces, but you’ll need to write an implementation based on some sort of webview in C# and then add initiateAccountLinking to user/Member. This is how accounts are linked in the app (since account linking requires browser logic like going to a specific URL and listening for a redirect that matches a certain regex). EDIT: Decided this does not need to be added in this iteration.

Point 1 is Completed.

**Comment by yuhanniu** [comment]
1. so there are three `docs` folders under core/user/tpp respectively that contain basically the same files. Is there a way to not duplicate those files? There is also a `.tmp` file in each of those `docs` folder which i assume is not needed.
2. could you rebase to master and pick up the latest the changes in master and make sure they are in the new SDK too? I will try to merge the code by the end of this week, thanks.

**Comment by mahendra-aps** [comment]
> Any update on this one? There are still merge conflicts

Done. Merge conflicts fixed.

**Comment by mahendra-aps** [comment]
> Moreover, the `Build()` method should return the corresponding TokenClient object. For example, `Tpp.TokenClient.Builder#Build()` should return `Tpp.TokenClient`. Right now all `Build()` methods in different modules return the same `Tokenio.TokenClient` and we cast it to the right type in the tests, this is not acceptable.

I think for now C# does not support Covariant Return Types like Java. So we did this way. Please let me know if you have any other way to do this, we can use that approach.

**Comment by yuhanniu** [comment]
I believe the [`new`](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/knowing-when-to-use-override-and-new-keywords) keyword allows you to override methods with different return types.
You are also not using generic type in the Builder class like you should.
Please we need you to try harder.
There is a test that is currently being skipped, is it being fixed?

**Comment by yuhanniu** [comment]
@JugrajSripalJain reminded me that test was brought up by you before and I forgot to look into it. The error is because the server is unable to verify the signature that is generated for the CancelToken operation. I suggest you carefully review the code for the signature generation and then compare the final Json that's being signed with the one that's generated by the Java sdk, and see if you can find anything suspicious.

**Comment by mahendra-aps** [comment]
> @JugrajSripalJain reminded me that test was brought up by you before and I forgot to look into it. The error is because the server is unable to verify the signature that is generated for the CancelToken operation. I suggest you carefully review the code for the signature generation, use the newest version the proto, and then compare the final Json that's being signed with the one that's generated by the Java sdk, and see if you can find anything suspicious.



> I believe the [`new`](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/knowing-when-to-use-override-and-new-keywords) keyword allows you to override methods with different return types.
> You are also not using generic type in the Builder class like you should.
> Please we need you to try harder.
> There is a test that is currently being skipped, is it being fixed?

@yuhanniu 
We are done with the changes suggested by you. Can you please have a look at those and let us know if anything else needs to be changed.
Updated the Builder class to a generic type.
Updated override methods to use different Signature type.
Also, we need your review on NewBuilder() method of Core Tokenclient class as we are getting an error here "Using the generic type 'TokenClient.Builder<T>' requires 1 type arguments". So, for now, we have removed this method in this commit as we don't see any references to this method. Is anyone going to use Core TokenClient directly?

**Comment by yuhanniu** [comment]
I have fixed the test, refer to the PR. I have't reviewed the NewBuilder method, will probably do that tomorrow.
Since we are building a new SDK, could you also remove all the methods that are marked as obsolete? I have removed some in my commit.

**Comment by mahendra-aps** [comment]
> I have fixed the test, refer to the [PR](https://github.com/tokenio/sdk-csharp/pull/50). I have't reviewed the NewBuilder method, will probably do that tomorrow.
> Since we are building a new SDK, could you also remove all the methods that are marked as obsolete? I have removed some in my commit.

Thanks, Yuhan. The test case is working fine on our side after this fix. We will start working on removing obsolete methods.

**Pull request merged**